cli/src/validate.rs

```diff
@@ -0,0 +1,261 @@
+use crate::constants::{EXIT_FAILURE, EXIT_SUCCESS};
+use crate::runner::build_filesets;
+use crate::scanner::{FileSet, FileSetCounter};
+use anyhow::Context;
+use colored::{ColoredString, Colorize};
+use console::Emoji;
+use context::junit::parser::{JunitParseError, JunitParser};
+use context::junit::validator::{
+    validate as validate_report, JunitReportValidation, JunitValidationLevel,
+};
+use quick_junit::Report;
+use std::io::BufReader;
+
+pub fn validate(junit_paths: Vec<String>, show_warnings: bool) -> anyhow::Result<i32> {
+    // scan files
+    let current_dir = std::env::current_dir()
+        .ok()
+        .and_then(|p| p.to_str().map(String::from))
+        .unwrap_or_default();
+    let (file_sets, file_counter) = build_filesets(&current_dir, &junit_paths, None, &None, None)?;
+    if file_counter.get_count() == 0 || file_sets.is_empty() {
+        return Err(anyhow::anyhow!("No JUnit files found to validate."));
+    }
+    print_matched_files(&file_sets, file_counter);
+
+    // parse and validate
+    let (reports, parse_errors) = parse_file_sets(file_sets)?;
+    if !parse_errors.is_empty() && show_warnings {
+        print_parse_errors(parse_errors);
+    }
+    let report_validations: Vec<(JunitReportValidation, String)> = reports
+        .into_iter()
+        .map(|report| (validate_report(&report.0), report.1))
+        .collect();
+
+    // print results
+    let (num_invalid_reports, num_suboptimal_reports) =
+        print_validation_errors(&report_validations);
+    if num_invalid_reports == 0 {
+        print_summary_success(report_validations.len(), num_suboptimal_reports);
+        Ok(EXIT_SUCCESS)
+    } else {
+        print_summary_failure(
+            report_validations.len(),
+            num_invalid_reports,
+            num_suboptimal_reports,
+        );
+        Ok(EXIT_FAILURE)
+    }
+}
+
+fn parse_file_sets(
+    file_sets: Vec<FileSet>,
+) -> anyhow::Result<(Vec<(Report, String)>, Vec<(JunitParseError, String)>)> {
```

Comment:
You can create a type alias for these more complex types and compose them together nicely for readability
https://doc.rust-lang.org/reference/items/type-aliases.html

For example:
```rs
type ParsedReportAndFilePaths = Vec<(Report, String)>;
type JunitParseErrorsAndFilePaths = Vec<(JunitParseError, String)>;
```

which would update your return type to:
```rs
anyhow::Result<(ParsedReportAndFilePaths, JunitParseErrorsAndFilePaths)>
```

Btw, it occurs to me that we could use a `BTreeMap` here though, no? Something like
```rs
type JunitFileToReportAndErrors = BTreeMap<String, (Vec<Report>, Vec<JunitParseError>)>;
```